### PR TITLE
[CNX-9823] add material quantity attribute to archicad elements

### PR DESCRIPTION
**Pull request by r-wessel**
Adds material quantities equivalent to the Revit output
Only measures:
- Beams
- Columns
- Mesh
- Morph
- Roof
- Shell
- Slab
- Wall
…elements.

### CODE REVIEW

File: ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp [hunk 1]
```diff
@@ -4,6 +4,373 @@
...
+		extendedQuantity.elemPartQuantities = isPartQuantity ? &elementPartQuantities : nullptr;
+		API_QuantityPar quantityParameters{};
+		quantityParameters.minOpeningSize = Eps;
+		MaterialQuantArray result;
+		if (auto error = ACAPI_Element_GetQuantities(element.header.guid, &quantityParameters, &extendedQuantity, &quantityMask); error != NoError)
+			return result;
+		if (isPartQuantity) {
+			if (elementPartQuantities.GetSize() > 0) {
+				elementQuantity.morph.baseLevel = elementPartQuantities[0].quantity.morph.baseLevel;
+				elementQuantity.morph.baseHeight = elementPartQuantities[0].quantity.morph.baseHeight;
```
Comment (JR-Morgan) [inline ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp:237]: If you don't mind tomorrow running me through this function a little

File: ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp [hunk 2]
```diff
@@ -4,6 +4,373 @@
...
+	 Collect the individual segments from a segmented assembly (e.g. beam/column)
+	 @param segments A pointer to the assembly segments
+	 @return An array containing the individual segments
+	 */
+	template<typename T, typename Insert>
+	auto getSegments(const T* segments, Insert insert) {
+		ElementArray result;
+		if (segments == nullptr)
+			return result;
+		for (auto segmentIndex = BMGetPtrSize((GSConstPtr) segments) / sizeof(T); segmentIndex--; ) {
```
Comment (JR-Morgan) [inline ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp:153]: same with this function...

**Review by JR-Morgan — CHANGES_REQUESTED** [review]

File: ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp [hunk 3]
```diff
@@ -4,6 +4,373 @@
...
+	auto measureQuantities(const API_Element& element, API_ElementQuantity& elementQuantity, API_Quantities& extendedQuantity,
+						   const API_QuantitiesMask& quantityMask) {
+		extendedQuantity.elements = &elementQuantity;
+		GS::Array<API_ElemPartQuantity> elementPartQuantities;
+		bool isPartQuantity = (element.header.type == API_MorphID);
+		extendedQuantity.elemPartQuantities = isPartQuantity ? &elementPartQuantities : nullptr;
+		API_QuantityPar quantityParameters{};
+		quantityParameters.minOpeningSize = Eps;
+		MaterialQuantArray result;
+		if (auto error = ACAPI_Element_GetQuantities(element.header.guid, &quantityParameters, &extendedQuantity, &quantityMask); error != NoError)
```
Comment (JR-Morgan) [inline ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp:232]: We have a fun CI error related to using `init if statements`...
Something about C++17, lets have a chat, see if its as simple as adding the suggested compiler flag.

https://app.circleci.com/pipelines/github/specklesystems/speckle-sharp/16007/workflows/d16c1f09-8f11-4e56-a9fd-63feecca93a5/jobs/148736

File: ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp [hunk 4]
```diff
@@ -164,19 +531,26 @@ GS::ErrCode SerializePropertyGroups (const GS::Array<API_PropertyDefinition>& de
...
 	if (!sendProperties && !sendListingParameters)
 		return NoError;
 
 	GS::Array<API_PropertyDefinition> elementDefinitions;
 	GS::Array < GS::Pair<API_ElemComponentID, GS::Array<API_PropertyDefinition>>> componentsDefinitions;
 
-	err = PropertyExportManager::GetInstance ()->GetElementDefinitions (element, sendProperties, sendListingParameters, systemItemPairs, elementDefinitions, componentsDefinitions);
+	GS::ErrCode err = PropertyExportManager::GetInstance ()->GetElementDefinitions (element, sendProperties, sendListingParameters, systemItemPairs, elementDefinitions, componentsDefinitions);
 	if (err != NoError)
-		return false;
```
[the remark this answers is not in the document]
Comment (r-wessel) [reply ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp:179]: I suspect this is a leftover from a function that originally returned bool and the author forgot to change it when it returned a GS::ErrCode. C++ makes implicit conversions between numeric types (including bool, where false = 0 and true = 1), so returning false would return 0 (= NoError), which indicates the call was successful. I can see no reason for returning a success value when an error occurred - it should return the error code.

[the remark this answers is not in the document]
Comment (JR-Morgan) [reply ConnectorArchicad/AddOn/Sources/AddOn/Commands/GetDataCommand.cpp:179]: I suspected as much

**Pull request merged**